File: Heap Sort/HeapSort.kt
package com.exam.kotlin.codingproject

/**
 * Created by msaycon on 09,Oct,2019
 */
class HeapSort {
    companion object{

        fun heapSort(numbersToSort: IntArray) : IntArray {
            val n = numbersToSort.size

            for (i in n / 2 - 1 downTo 0){
                generateHeap(numbersToSort, n, i)
            }

            for (i in n - 1 downTo 0) {
                // Move current root to end
                val temp = numbersToSort[0]
                numbersToSort[0] = numbersToSort[i]
                numbersToSort[i] = temp

                generateHeap(numbersToSort, i, 0)
            }

            return numbersToSort
        }

        private fun generateHeap(arr: IntArray, n: Int, i: Int) {
            var largest = i // Initialize largest as root
            val l = 2 * i + 1 // left = 2*i + 1
            val r = 2 * i + 2 // right = 2*i + 2

            // If left child is larger than root
            if (l < n && arr[l] > arr[largest])
                largest = l

            // If right child is larger than largest so far
            if (r < n && arr[r] > arr[largest])
                largest = r

            // If largest is not root
            if (largest != i) {
                val swap = arr[i]
                arr[i] = arr[largest]
                arr[largest] = swap

                generateHeap(arr, n, largest)
            }
        }

    }
}

File: Heap Sort/heapSort.c
#include<stdio.h>
#include<conio.h>

//swap function to swap values
void swap(int *n,int *p)
{
    int temp=*n;
    *n=*p;
    *p=temp;
}
// To heapify a subtree rooted with node i which is
// an index in arr[]. n is size of heap
void heapify(int a[],int n,int i)
{
    int lar_ind=i;
    int l=2*i+1;
    int r=2*i+2;
    // If left child is larger than root
    if(l<n && a[l]>a[lar_ind])
    {
        lar_ind=l;
    }
    // If right child is larger than largest so far
    if(r<n && a[r]>a[lar_ind])
    {
        lar_ind=r;
    }
    // If largest is not root
    if(lar_ind!=i)
    {
        swap(&a[i],&a[lar_ind]);
        // Recursively heapify the affected sub-tree
        heapify(a,n,lar_ind);
    }
}
void heapsort(int a[],int n)
{
    // Build heap (rearrange array)
    for(int i=n/2-1;i>=0;i--)
    {
        heapify(a,n,i);
    }
    // One by one extract an element from heap
    for(int i=n-1;i>0;i--)
    {
        // Move current root to end
        swap(&a[0],&a[i]);
        // call max heapify on the reduced heap
        heapify(a,i,0);
    }
}
// Driver program
void main()
{
    int n;
    printf("enter NO. of elements");
    scanf("%d",&n);
    int arr[n];
    for(int i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    heapsort(arr,n);
    printf("sorted array by heapsort is");
    for(int i=0;i<n;i++)
    {
        printf("%d\n",arr[i]);
    }


}
